<a href="https://colab.research.google.com/github/adichat08/Support-Vector-Classifier-for-Predicting-Survival-Likelihood-of-Hepatitis-Patients/blob/main/Adding_Percent_Confidence_to_the_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Adding Percent Confidence to the Model



At this point, the model simply returns a prediction on whether or not it thinks that the patient is likely to die. The objective in this subsection is to add percent confidence to the model's output. The model already holds data in the decision function that can allow for such a metric. This will greatly help anyone looking at results from the model understand a prediction more clearly.

In [ ]:
# printing decision certainties for the SVC
print('SVC decision certainty:\n{}'.format(svc.decision_function(X_test)))
# printing decision certainties for the SVC
print('SVC decision certainty maximum:\n{}'.format(np.max(svc.decision_function(X_test))))

SVC decision certainty:
[ 2.97694625  2.16968269  1.91178736 -0.48663314  2.48874233  3.45440417
  2.62581712 -2.34392841  1.6424457   4.05941997  2.31283638  3.21961146
  0.37281569  2.53935946  0.09120863  3.5636346   2.38709412  2.57396344
  3.53912598  1.76732509  0.2831205   2.35099337  1.58191222 -1.48675427
  2.47396667  4.22793795  0.56968956  1.83294886  3.00889823  0.70965508]
SVC decision certainty maximum:
4.2279379483652715


At this point, an arbitrary value can be chosen to represent maximum confidence. Since 4.23 appears to be the decision function value for the most confident prediction, 4.5 is likely to be a fitting choice for a 100% confident. Obviously, 100% confidence doesn't make it any more right than 10% confidence, because it is possible for the model to be very confident in an inaccurate prediction(for a bad model, which this doesn't appear to be).

In [ ]:
class Modified_SVC(BaseEstimator, ClassifierMixin):
  def __init__(self, C = 100, gamma = 0.01, kernel = 'sigmoid', random_state = 42):
    self.C = C
    self.gamma = gamma
    self.kernel = kernel
    self.random_state = random_state

  def fit(self, X, y=None):
    self.X = X
    self.y = y
    from sklearn.svm import SVC
    global svc
    svc = SVC(C = 100, gamma = 0.01, kernel = 'sigmoid', random_state = 42)
    svc.fit(X,y)
    return self

  def predict(self, X):
    self.X = X
    prediction = svc.predict(self.X)
    svc_decision_function = svc.decision_function(X_test)
    confidence_lst = []
    for i in svc_decision_function:
      confidence_lst.append((np.abs(i)/np.max(svc_decision_function))*100)
    prediction_confidence_dict = {"Prediction": list(prediction), "Percent Confidence in Prediction": confidence_lst}
    prediction_confidence_df = pd.DataFrame(prediction_confidence_dict)
    return prediction_confidence_df


modified_svc = Modified_SVC()
modified_svc.fit(X_train,y_train)
modified_svc.predict(X_test)

Prediction  Percent Confidence in Prediction
0            2                         70.411304
1            2                         51.317752
2            2                         45.217962
3            1                         11.509940
4            2                         58.864211
5            2                         81.704231
6            2                         62.106331
7            1                         55.439045
8            2                         38.847441
9            2                         96.014180
10           2                         54.703650
11           2                         76.150868
12           2                          8.817908
13           2                         60.061418
14           2                          2.157284
15           2                         84.287770
16           2                         56.460008
17           2                         60.879877
18           2                         83.708087
19           2                         41.801112
20           2                          6.696420
21           2                         55.606147
22           2                         37.415691
23           1                         35.164997
24           2                         58.514735
25           2                        100.000000
26           2                         13.474407
27           2                         43.353258
28           2                         71.167038
29           2                         16.784898

In [ ]:
modified_svc.predict(X_test)['Prediction']

0     2
1     2
2     2
3     1
4     2
5     2
6     2
7     1
8     2
9     2
10    2
11    2
12    2
13    2
14    2
15    2
16    2
17    2
18    2
19    2
20    2
21    2
22    2
23    1
24    2
25    2
26    2
27    2
28    2
29    2
Name: Prediction, dtype: int64

In [ ]:
class Modified_SVC_1(BaseEstimator, ClassifierMixin):
  def __init__(self, C = 100, gamma = 0.01, kernel = 'sigmoid', random_state = 42):
    self.C = C
    self.gamma = gamma
    self.kernel = kernel
    self.random_state = random_state

  def fit(self, X, y=None):
    self.X = X
    self.y = y
    from sklearn.svm import SVC
    global svc
    svc = SVC(C = 100, gamma = 0.01, kernel = 'sigmoid', random_state = 42)
    svc.fit(X,y)
    return self

  def predict(self, X):
    self.X = X
    prediction = svc.predict(self.X)
    svc_decision_function = svc.decision_function(X_test)
    # confidence_lst = []
    # for i in svc_decision_function:
    #   confidence_lst.append((np.abs(i)/4.2279379483652715)*100)
    if prediction[0] == 1:
      pred_conf = f'The patient has a {round(100-((np.abs(i)/29.5)*100))}% chance of survival'
    if prediction[0] == 2:
      pred_conf = f'The patient has a {round((np.abs(i)/29.5)*100)}% chance of survival'

    return pred_conf


modified_svc_1 = Modified_SVC_1()
modified_svc_1.fit(X_train,y_train)
modified_svc_1.predict(X_test)

'The patient has a 98% chance of survival'